File: Data Structures & Algorithms/products-of-array-discluding-self/submission-3.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        n = len(nums)
        answer = [1]*n
        current = nums[0]
        for i in range(1,n):
            answer[i] = answer[i]*current
            current *= nums[i]
        
        current = nums[n-1]
        for i in range(n-2, -1, -1):
            answer[i] = answer[i]*current
            current *= nums[i]
        
        return answer

File: Data Structures & Algorithms/top-k-elements-in-list/submission-3.cpp
class Solution {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        vector<int> answer;
        unordered_map<int, int> m;
        for (int i : nums){
            m[i]++;
        }

        int n = nums.size();

        vector<vector<int>> bucket(n + 1, vector<int>());
        for (auto p : m){
            bucket[p.second].push_back(p.first);
        }

        for (int i = bucket.size()-1; i >= 0 && k > 0;i--){
            while (!bucket[i].empty() && k > 0){
                answer.push_back(bucket[i].back());
                bucket[i].pop_back();
                k--;
            }
        }

        return answer;
    }
};
